torch/csrc/jit/interpreter.cpp

```diff
@@ -661,173 +685,239 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
     // stack when we suspend, record where it starts so we return the right
     // stack
     if (stack_start_ == -1) {
-      TORCH_INTERNAL_ASSERT(stack.size() >= function->n_inputs);
-      stack_start_ = stack.size() - function->n_inputs;
+      TORCH_INTERNAL_ASSERT(stack.size() >= frames.back().function->n_inputs);
+      stack_start_ = stack.size() - frames.back().function->n_inputs;
     } else {
       // during restarts, all of the stack is always our own, so we leave
       // nothing
       stack_start_ = 0;
     }
+
+    ActiveFrame af(frames.back());
     try {
-      return runInterpreterLoop(stack);
+      while (true) {
+        // std::cout << "RUNNING ";
+        // frames.back().function->dump(std::cout, af.pc);
+        Instruction inst = af.instructions[af.pc];
+        switch (inst.op) {
+          case OP:
+            af.operators[inst.X](stack);
+            ++af.pc;
+            break;
+          case LOAD:
+            stack.emplace_back(reg(inst.X));
+            ++af.pc;
+            break;
+          case MOVE:
+            stack.emplace_back(std::move(reg(inst.X)));
+            ++af.pc;
+            break;
+          case STORE:
+            reg(inst.X) = pop(stack);
+            ++af.pc;
+            break;
+          case STOREN:
+            for (size_t i = inst.N; i > 0; --i) {
+              reg(inst.X + i - 1) = pop(stack);
+            }
+            ++af.pc;
+            break;
+          case DROP:
+            pop(stack);
+            ++af.pc;
+            break;
+          case DROPR:
+            reg(inst.X) = IValue();
+            ++af.pc;
+            break;
+          case LOADC:
+            stack.emplace_back(af.constants[inst.X]);
+            ++af.pc;
+            break;
+          case JF:
+            af.pc += (pop(stack).toBool()) ? 1 : inst.X;
+            break;
+          case JMP:
+            af.pc += inst.X;
+            break;
+          case LOOP: {
+            // stack: iteration_count, max_iter, cond, loop_carried_deps...
+            auto frame = stack.end() - (inst.N + 1);
+            int64_t trip_count = frame[0].toInt();
+            int64_t max_trip_count = frame[1].toInt();
+            bool cond = frame[2].toBool();
+            if (trip_count < max_trip_count && cond) {
+              frame[2] = trip_count;
+              frame[0] = trip_count + 1;
+              ++af.pc;
+            } else {
+              size_t n_loop_carried = inst.N - 2;
+              for (size_t i = 0; i < n_loop_carried; ++i) {
+                frame[i] = std::move(frame[i + 3]);
+              }
+              drop(stack, 3); // iteration_count, max_iter, cond
+              af.pc += inst.X;
+            }
+          } break;
+          case CALL: {
+            const Code& code =
+                af.functions[inst.X]->get_executor().getPlanFor(stack).code;
+            frames.back().pc = af.pc + 1;
+            enterFrame(code);
```

Comment:
I think `enterFrame` and `leaveFrame` should set the active frame. That would require `af` to be a member of the interpreter state though.